data/totem-audio-preview.1
.\" Copyright (C) 2014 Milan Kupcevic <milan@physics.harvard.edu>
.\"
.\" This file is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This file is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License 
.\" along with this; if not write to the Free Software Foundation, Inc.
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
.TH totem\-audio\-preview 1 "2014\-06\-08" "GNOME" "User Commands"
.SH NAME
totem\-audio\-preview \- play audio files
.SH SYNOPSIS
.B totem\-audio\-preview
.RI [ OPTION ]... " " [ FILE ]...
.SH DESCRIPTION
.B totem\-audio\-preview 
is primarily used by GNOME applications for audio preview purposes.
.SH OPTIONS
.TP
.B \-h, \-\-help
Display help.
.TP
.B \-\-help-gst
Display GStreamer options.
.TP
.B \-m, \-\-mimetype
Display supported mime\-types.
.TP
.B \-\-g\-fatal\-warnings
Make all warnings fatal.
.PP
When no FILE is specified, read standard input.
.SH AUTHOR
.B totem\-audio\-preview 
was written by Bastien Nocera <hadess@hadess.net>.
.PP 
This manual page was written by Milan Kupcevic for the Debian Project.
.SH SEE ALSO
.PP
.BR \%totem\-video\-thumbnailer (1),
.BR \%totem (1), 
.BR \%nautilus (1)


NEWS
New features and significant updates in version...

Major changes in 3.18.1:
- Make DVD menu navigation work again
- Updated translations

Major changes in 3.18.0:
- Fix rendering bugs in the search entry

Major changes in 3.17.92:
- None compared to 3.16.4

Major changes in 3.16.4:
- Fix selection not working after visiting Channels page
- Fix MPRIS plugin following internal API changes
- Revert Wayland compatibility changes, it's not a required
  change in GNOME 3.16

Major changes in 3.16.3:
- Fix ASF videos not being thumbnailed
- Bring back video inside the window under Wayland
- Fix Skip To dialogue not working
- Fix thumbnailer with newer versions of gstreamer-vaapi installed
- Update YouTube API key

Major changes in 3.16.2:
- Fix popup not showing up on touchscreens
- Fix potential crash on 32-bit systems

Major changes in 3.16.1:
- Loads of performance fixes, related to thumbnailing
  when a video is playing
- Fix (again) doubling files added on the command-line
- Don't monitor the tracker source, it adds incomplete
  items to the view
- Fix Videos icon not showing in volume control
- Provide a symbolic variant of the app icon
- Fix frames not getting applied for new thumbnails
- Don't warn when video rotation is unset

Major changes in 3.16.0:
- Translation updates

Major changes in 3.15.92:
- Update thumbnail styling
- Fix a crash on exit
- Fix application icon not showing in volume control
- Disable popover transitions for video menus
- Fix downloading subtitles with newer Python versions
- Translation updates

Major changes in 3.15.91:
- Build fixes
- Internationalisation fixes
- Update appdata file
- Translation updates

Major changes in 3.15.90:
- Better missing plugins support with GNOME Software
- Modify appearance of source icons
- Make view switcher button homogeneous
- Use intltool to translate GSettings schema files directly
- Better translator comment for subtitle-encoding
- Remove outdated helper script, which used Python 2

Major changes in 3.15.4:
- Remove spacing around preferences tabs
- Stop thumbnailing when exiting, and when playback starts
- Handle "Powerpoint" buttons on remotes
- Keep the controls visible when seeking with the slider
- Show controls when seeking with the keyboard
- Use the theme's foreground color to render the symbolics
- Only show videos longer than 5 seconds in the overview
- Hide source which provide torrent URLs (for now)
- Remove use of a number of deprecated properties
- Require a newer cairo for hi-dpi support
- Fix Totem doubling files added on the cmd line
- Fix missing translation support in thumbnailer/gallery support

Major changes in 3.14.1:
- Fix lingering white line in fullscreen
- Fix bug in MPRIS support when used via Rygel
- Thumbnail bookmarked movies as well
- Fix starting Totem via D-Bus (gnome-open)
- Port from appdata-tools to appstream-glib
- Add a keyboard shortcut for starting subtitles download

Major changes in 3.14.0:
- save-file: Handle an unset Videos directory
- Fix popup disappearing when volume is being set
- Remove border from error dialogue
- Require newer versions of clutter-gtk and gst-plugins-base

Major changes in 3.13.92:
- Fix opening files following the D-Bus-ification
- libgd updates to fix the appearance of check boxes and subtitle labels

Major changes in 3.13.91:
- Generate thumbnails for the videos in the main view when they're missing
- Rename totem.desktop to org.gnome.Totem.desktop
- Updated translations

Major changes in 3.13.90:
- Rename "Recent" section to "Videos"
- Remove visualisation support
- Remove browser plugin
- Better error message for RTSP blocked by firewall
- Automatically rotate mobile phone videos
- Allow seeking past fill-level
- Postpone loading library after playback
- Fix headerbar title when playing search results
- Fix crasher on startup with newer GTK+
- Fix crash when browsing Euronews source
- Use correct icon name for "gear" menu
- Don't allow adding duplicate items to recent view
- User new "audio-filter" property
- Fix loading Audio CDs' Wave files
- Fix potential data loss when removing multiple files
- Reset loaded text subtitles when closing video
- Fix hidden cursor with visible popup
- Move the "un-fullscreen" button to the corner

OpenSubtitles:
- Don't throw an error when the language is unknown
- Fix subtitles suffix when saving it
- Add support for Basque

Major changes in 3.12.1:
- Bump required versions of grilo and GStreamer
- Fix delete button's initial state
- Switch to the player view quicker
- Add support for starting a playlist paused
- Escape selection mode when pressing Esc
- Fix shuffled items starting to play in random order
- Hide popups when leaving the window, but not when a popover
  (volume, menus) is visible
- Don't toggle popups on touchpad clicks, just on touchscreens
- Fix scroll-wheel going forward 2 minutes instead of 1
- Save the downloaded subtitles in the video directory
- Re-add drag'n'drop support for the video widget

Major changes in 3.12.0:
- Fix artifacts caused by GtkPopovers
- Don't add containers in the Recent view
- Various opensubtitles plugin bug fixes
- Translation updates

Major changes in 3.11.91:
- Use popovers for menus, and remove their shadows
- Hide popups after 2 seconds not 5
- Fix crashers when deleting items
- Only resolve using the passed source
- Remove stock icons usage
- Don't offer to save file in ~/Videos subdirs
- Fix empty window title in some cases
- Fix handling of kbd shortcuts with multiple modifiers
- Blacklist adult sources for now
- Fix ngettext usage
- Avoid queueing multiple thumbnail updates
- Add names to timeouts and idles
- Don't monitor the filesystem plugin
- Translation updates

Major changes in 3.11.90:
- Add sideways swipe gestures to seek inside the window
- Move Open/Add menu items to the "Add" button
- Add drop support to the "Recent" section of the overview
- Implement "Make Available Offline"
- Maximise window by default
- Make F10 show the gear menu

src/totem-audio-preview.c
/* 
 * Copyright (C) 2006 Bastien Nocera <hadess@hadess.net>
 * Copyright (C) 2007 Bastien Nocera <hadess@hadess.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA.
 *
 * The Totem project hereby grant permission for non-gpl compatible GStreamer
 * plugins to be used and distributed together with GStreamer and Totem. This
 * permission are above and beyond the permissions granted by the GPL license
 * Totem is covered by.
 *
 * Monday 7th February 2005: Christian Schaller: Add exception clause.
 * See license_change file for details.
 *
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <glib.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <glib/gi18n.h>
#include <gio/gio.h>

#include "totem-gst-helpers.h"
#include "totem-resources.h"
#define WANT_AUDIO_MIME_TYPES 1
#include "totem-mime-types.h"

static gboolean show_mimetype = FALSE;
static gboolean g_fatal_warnings = FALSE;
static char **filenames = NULL;

static void
print_mimetypes (void)
{
	guint i;

	for (i =0; audio_mime_types[i] != NULL; i++) {
		g_print ("%s\n", audio_mime_types[i]);
	}
}

static const GOptionEntry entries[] = {
	{"mimetype", 'm', 0, G_OPTION_ARG_NONE, &show_mimetype, "List the supported mime-types", NULL},
	{"g-fatal-warnings", 0, 0, G_OPTION_ARG_NONE, &g_fatal_warnings, "Make all warnings fatal", NULL},
	{G_OPTION_REMAINING, '\0', 0, G_OPTION_ARG_FILENAME_ARRAY, &filenames, "Audio files to play back", NULL},
	{NULL}
};

static void
setup_play (GstElement *play)
{
	GstElement *audio_sink;
	audio_sink = gst_element_factory_make ("autoaudiosink", "audio-sink");
	g_object_set (play,
		      "audio-sink", audio_sink,
		      "flags", GST_PLAY_FLAG_SOFT_VOLUME | GST_PLAY_FLAG_AUDIO,
		      NULL);
}

static GstBusSyncReply
error_handler (GstBus *bus,
	       GstMessage *message,
	       GstElement *play)
{
	GstMessageType msg_type;

	msg_type = GST_MESSAGE_TYPE (message);
	switch (msg_type) {
	case GST_MESSAGE_ERROR:
		totem_gst_message_print (message, play, "totem-audio-preview-error");
		exit (1);
	case GST_MESSAGE_EOS:
		exit (0);

	case GST_MESSAGE_ASYNC_DONE:
	case GST_MESSAGE_UNKNOWN:
	case GST_MESSAGE_WARNING:
	case GST_MESSAGE_INFO:
	case GST_MESSAGE_TAG:
	case GST_MESSAGE_BUFFERING:
	case GST_MESSAGE_STATE_CHANGED:
	case GST_MESSAGE_STATE_DIRTY:
	case GST_MESSAGE_STEP_DONE:
	case GST_MESSAGE_CLOCK_PROVIDE:
	case GST_MESSAGE_CLOCK_LOST:
	case GST_MESSAGE_NEW_CLOCK:
	case GST_MESSAGE_STRUCTURE_CHANGE:
	case GST_MESSAGE_STREAM_STATUS:
	case GST_MESSAGE_APPLICATION:
	case GST_MESSAGE_ELEMENT:
	case GST_MESSAGE_SEGMENT_START:
	case GST_MESSAGE_SEGMENT_DONE:
	case GST_MESSAGE_DURATION_CHANGED:
	case GST_MESSAGE_LATENCY:
	case GST_MESSAGE_ASYNC_START:
	case GST_MESSAGE_REQUEST_STATE:
	case GST_MESSAGE_STEP_START:
	case GST_MESSAGE_QOS:
	case GST_MESSAGE_PROGRESS:
	case GST_MESSAGE_TOC:
	case GST_MESSAGE_RESET_TIME:
	case GST_MESSAGE_STREAM_START:
	case GST_MESSAGE_ANY:
	case GST_MESSAGE_NEED_CONTEXT:
	case GST_MESSAGE_HAVE_CONTEXT:
	default:
		/* Ignored */
		;;
	}

	return GST_BUS_PASS;
}

static void
setup_errors (GstElement *play)
{
	GstBus *bus;

	bus = gst_element_get_bus (play);
	gst_bus_set_sync_handler (bus, (GstBusSyncHandler) error_handler, play, NULL);
}

static void
setup_filename (GstElement *play)
{
	if (filenames != NULL) {
		GFile *file;
		char *uri;

		file = g_file_new_for_commandline_arg (filenames[0]);
		uri = g_file_get_uri (file);
		g_object_unref (file);

		g_object_set (play, "uri", uri, NULL);
		g_free (uri);
	} else {
		g_object_set (play, "uri", "fd://0", NULL);
	}
}

int main (int argc, char **argv)
{
	GOptionGroup *options;
	GOptionContext *context;
	GError *error = NULL;
	GMainLoop *loop;
	GstElement *play;

	bindtextdomain (GETTEXT_PACKAGE, GNOMELOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	g_set_application_name (_("Audio Preview"));
	g_setenv("PULSE_PROP_application.icon_name", "totem", TRUE);
	g_setenv("PULSE_PROP_media.role", "music", TRUE);

	context = g_option_context_new ("Plays audio passed on the standard input or the filename passed on the command-line");
	options = gst_init_get_option_group ();
	g_option_context_add_main_entries (context, entries, GETTEXT_PACKAGE);
	g_option_context_add_group (context, options);

	if (g_option_context_parse (context, &argc, &argv, &error) == FALSE) {
		g_print ("couldn't parse command-line options: %s\n", error->message);
		g_error_free (error);
		return 1;
	}

	if (g_fatal_warnings) {
		GLogLevelFlags fatal_mask;

		fatal_mask = g_log_set_always_fatal (G_LOG_FATAL_MASK);
		fatal_mask |= G_LOG_LEVEL_WARNING | G_LOG_LEVEL_CRITICAL;
		g_log_set_always_fatal (fatal_mask);
	}

	if (show_mimetype == TRUE) {
		print_mimetypes ();
		return 0;
	}

	if (filenames != NULL && g_strv_length (filenames) != 1) {
		char *help;
		help = g_option_context_get_help (context, FALSE, NULL);
		g_print ("%s", help);
		g_free (help);
		return 1;
	}

	play = gst_element_factory_make ("playbin", "play");
	setup_play (play);
	setup_filename (play);
	setup_errors (play);
	totem_resources_monitor_start (NULL, -1);
	gst_element_set_state (play, GST_STATE_PLAYING);

	loop = g_main_loop_new (NULL, TRUE);
	g_main_loop_run (loop);

	return 0;
}

